### PR TITLE
dev/core#4537 Ensure that Event Registration email works when CiviCon…

### PR DESCRIPTION
**Pull request by seamuslee001**
…tribute component is disabled and add unit test

Overview
----------------------------------------
This fixes an issue where performing an event registration without the CiviContribute Component / Extension enabled errors because LineItem APIv4 class isn't present and also fixes issues with smarty template rendering in similar

Before
----------------------------------------
Event Registration fails without CiviContribute enabled

After
----------------------------------------
Works

ping @eileenmcnaughton @totten @MegaphoneJon

### CODE REVIEW

File: Civi/Token/TokenCompatSubscriber.php [hunk 1]
```diff
@@ -56,7 +56,10 @@ public function setupSmartyAliases(TokenValueEvent $e) {
    */
   public function onRender(TokenRenderEvent $e): void {
     $useSmarty = !empty($e->context['smarty']);
-    $e->string = $e->getTokenProcessor()->visitTokens($e->string, function() {
+    $e->string = $e->getTokenProcessor()->visitTokens($e->string, function($token = NULL, $entity = NULL, $field = NULL, $filterParams = NULL) {
```
Comment (seamuslee001) [inline Civi/Token/TokenCompatSubscriber.php:59]: without CiviContribute there were issues in the event online receipt (and other related templates) because they would have tokens like 

{contribution.tax_amount|boolean} in if statements and this was filtering that down to an empty string which then if there were multiple parts to the if e.g. {if $showTaxAmount && {contribution.tax_amount|boolean}} it was being passed to smarty as {if $showTaxAmount && } which of course is a PHP syntax error this causes it to resolve any that use boolean filter to be 0

File: tests/phpunit/CRM/Event/Form/Registration/ConfirmTest.php [hunk 2]
```diff
@@ -721,4 +721,37 @@ protected function submitPaidEvent(array $submitValues = []): void {
...
+  public function testRegistrationWithoutCiviContributeEnabled(): void {
+    $mut = new CiviMailUtils($this, TRUE);
+    $event = $this->eventCreateUnpaid([
+      'has_waitlist' => 1,
+      'max_participants' => 1,
+      'start_date' => 20351021,
+      'end_date' => 20351023,
+      'registration_end_date' => 20351015,
+    ]);
+    CRM_Core_BAO_ConfigSetting::disableComponent('CiviContribute');
```
Comment (seamuslee001) [inline tests/phpunit/CRM/Event/Form/Registration/ConfirmTest.php:733]: If I moved this before eventCreteUnpaid I ran into a whole bunch of issues with Example Data creation so decided to skip overall of that for now

File: CRM/Event/WorkflowMessage/ParticipantTrait.php [hunk 3]
```diff
@@ -67,6 +67,10 @@ trait CRM_Event_WorkflowMessage_ParticipantTrait {
    */
   protected $participantContacts;
 
+  private function isContributionComponentEnabled(): bool {
+    return CRM_Core_Component::isEnabled('CiviContribute');
```
Comment (eileenmcnaughton) [inline CRM/Event/WorkflowMessage/ParticipantTrait.php:71]: So we did switch to checking for entity present in tokens - ie https://github.com/civicrm/civicrm-core/pull/27201/files#diff-1577dbbf2e3373be5f6fb29637b0ec2bee12390aabb24997d951b3509e44b7f3R61

perhaps that is more important in tokens as loaded earlier

File: Civi/Token/TokenCompatSubscriber.php (hunk 1 above)
Comment (eileenmcnaughton) [reply Civi/Token/TokenCompatSubscriber.php:59]: OK - let me see if I can move this into where booleans are otherwise rendered - ie this sort of area

https://github.com/civicrm/civicrm-core/blob/1ce5c9bf6b5d93e4ffd6bc2f44d5226aeb1141d2/Civi/Token/TokenProcessor.php#L494-L499

File: CRM/Event/WorkflowMessage/ParticipantTrait.php (hunk 3 above)
Comment (seamuslee001) [reply CRM/Event/WorkflowMessage/ParticipantTrait.php:71]: Just to be clear the action object provider is a known list of APIv4 entities it isn't specific about if an entity is in tokens or not but I suppose I could switch this to that style if we wanted it like that

File: Civi/Token/TokenCompatSubscriber.php (hunk 1 above)
Comment (seamuslee001) [reply Civi/Token/TokenCompatSubscriber.php:59]: So the problem is we don't get to there because of https://github.com/civicrm/civicrm-core/blob/1ce5c9bf6b5d93e4ffd6bc2f44d5226aeb1141d2/Civi/Token/TokenProcessor.php#L381

File: CRM/Event/WorkflowMessage/ParticipantTrait.php (hunk 3 above)
Comment (colemanw) [reply CRM/Event/WorkflowMessage/ParticipantTrait.php:71]: IMO it's best to get away from using `CRM_Core_Component` as much as possible; as future-proofing and to make it easier to eventually get rid of it.

Comment (seamuslee001) [reply CRM/Event/WorkflowMessage/ParticipantTrait.php]: ok I have updated now and also renamed function to remove the word Component as well

File: Civi/Token/TokenCompatSubscriber.php (hunk 1 above)
Comment (eileenmcnaughton) [reply Civi/Token/TokenCompatSubscriber.php:59]: @seamuslee001 yeah we could do 

image


Which feels more right to me - in that generally token filters are handled in the token system. OTOH I am somewhat terrified of the code in `vistiTokens` and associated functions cos it's hard to follow which might be a case for not touching it

Comment (seamuslee001) [reply Civi/Token/TokenCompatSubscriber.php:59]: yeh that might be right but yeh likewise and it seemed like as this was doing the "cleanup" it might be the most sensible place to put things

File: CRM/Event/WorkflowMessage/ParticipantTrait.php [hunk 4]
```diff
@@ -67,6 +67,10 @@ trait CRM_Event_WorkflowMessage_ParticipantTrait {
    */
   protected $participantContacts;
 
+  private function isCiviContributionEnabled(): bool {
```
Comment (eileenmcnaughton) [inline CRM/Event/WorkflowMessage/ParticipantTrait.php:70]: not blocking but `isCiviContributeEnabled()` feels slightly more correct

File: Civi/Token/TokenCompatSubscriber.php (hunk 1 above)
Comment (eileenmcnaughton) [reply Civi/Token/TokenCompatSubscriber.php:59]: ok - well I'm otherwise happy with this PR so @totten can merge this as is or opt for the other location to handle unset booleans

File: CRM/Event/WorkflowMessage/ParticipantTrait.php (hunk 4 above)
Comment (seamuslee001) [reply CRM/Event/WorkflowMessage/ParticipantTrait.php]: ok renamed now

**Comment by eileenmcnaughton** [comment]
Happy leap year

File: Civi/Token/TokenCompatSubscriber.php (hunk 1 above)
Comment (totten) [reply Civi/Token/TokenCompatSubscriber.php:59]: OK, this is the first time I've really looked at `|boolean`. So I'm trying to catch-up. Here's what I'm understanding:

* The idea in 6543dfe601b0c5ac58b2062d8fe5d409c1a9d313 is to help write Smarty conditionals that key-off of token-data, eg
    ```smarty
    {if {contact.first_name|boolean}}...{/if}
    ```
* This kind of `{if {token}}` notation is the quintessential quirk of Token-Smarty format. It's why Token-Smarty is evil and should never be used. It's also why we have to keep it around -- because changing to a simpler/singular language would break these conditionals.
* So I guess the idea of `|boolean` is to make the conditional a bit safer, eg
    ```smarty
    {if {contact.first_name}}...{/if}        {* More evil *}
    {if {contact.first_name|boolean}}...{/if} {* Less evil *}
    ```
* In the callpath of this specific bug, we have a CiviEvent message with expression:
    ```smarty
    {if $isShowTax && {contribution.tax_amount|boolean}}
    ```
* In CiviEvent contexts, the token `{contribution.tax_amount}` may or may not be defined. (*CiviEvent is only softly dependent on CiviContribute.*)
* The approach here is to define `{any_unkn.own|boolean}` to output `0`.
* The question you're mulling is: Where, exactly, should we put in the override to make `{any_unkn.own|boolean}` evaluate to `0`?

Comment (totten) [reply Civi/Token/TokenCompatSubscriber.php:59]: So, yes, to my eye, both locations work, and the `TokenProcessor::$getToken()`  seems a bit more poignant than `TokenCompatSubscriber::onRender()`.

Whichever location it's in, I think it should have a code-comment explaining why it's there, e.g.

```php
if ($filterParams && $filterParams[0] === 'boolean') {
  // This token was missed during primary rendering, and it's supposed to be coerced to boolean.
  // Treat an unknown token as false-y.
  return 0;
}
```

Comment (eileenmcnaughton) [reply Civi/Token/TokenCompatSubscriber.php:59]: @totten that seems about right

Comment (totten) [reply Civi/Token/TokenCompatSubscriber.php:59]: Force-pushed a commit with the extra comment.
